File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s:
            return ""

        start, end = 0, 0

        def expandAroundCenter(left, right):
            while left >= 0 and right < len(s) and s[left] == s[right]:
                left -= 1
                right += 1
            return right - left - 1  # Subtract 1 at the end because left and right are beyond the palindrome boundaries

        for i in range(len(s)):
            len1 = expandAroundCenter(i, i)  # Check for odd-length palindromes
            len2 = expandAroundCenter(i, i + 1)  # Check for even-length palindromes

            # Use the longer one
            length = max(len1, len2)

            if length > end - start:
                start = i - (length - 1) // 2
                end = i + length // 2

        return s[start:end + 1]

File: 5-longest-palindromic-substring/longest-palindromic-substring.cpp
#include <string>
using std::string;

class Solution {
public:
    string longestPalindrome(string s) {
        if (s.empty()) return "";
        if (s.size() == 1) return s;

        int start = 0, end = 0;

        for (int i = 0; i < s.size(); i++) {
            // Check for odd-length palindromes
            int len1 = expandAroundCenter(s, i, i);
            // Check for even-length palindromes
            int len2 = expandAroundCenter(s, i, i + 1);

            // Find the longest of the two
            int len = std::max(len1, len2);

            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }

        return s.substr(start, end - start + 1);
    }

private:
    int expandAroundCenter(const string& s, int left, int right) {
        while (left >= 0 && right < s.size() && s[left] == s[right]) {
            left--;
            right++;
        }
        // Subtract 1 at the end because when the while loop exits,
        // left and right are one step beyond the palindrome boundaries
        return right - left - 1;
    }
};
